multiapps-controller-web/src/main/java/org/cloudfoundry/multiapps/controller/web/upload/AsyncUploadJobExecutor.java

```diff
@@ -0,0 +1,365 @@
+import org.cloudfoundry.multiapps.controller.persistence.model.FileEntry;
+import org.cloudfoundry.multiapps.controller.persistence.model.ImmutableAsyncUploadJobEntry;
+import org.cloudfoundry.multiapps.controller.persistence.model.ImmutableFileEntry;
+import org.cloudfoundry.multiapps.controller.persistence.services.AsyncUploadJobService;
+import org.cloudfoundry.multiapps.controller.persistence.services.FileService;
+import org.cloudfoundry.multiapps.controller.web.Constants;
+import org.cloudfoundry.multiapps.controller.web.Messages;
+import org.cloudfoundry.multiapps.controller.web.util.SecurityContextUtil;
+import org.cloudfoundry.multiapps.mta.handlers.ArchiveHandler;
+import org.cloudfoundry.multiapps.mta.handlers.DescriptorParserFacade;
+import org.cloudfoundry.multiapps.mta.model.DeploymentDescriptor;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.http.HttpHeaders;
+import org.springframework.http.HttpStatus;
+
+@Named
+public class AsyncUploadJobExecutor {
+
+    private static final int INPUT_STREAM_BUFFER_SIZE = 16 * 1024;
+
+    private static final Duration HTTP_CONNECT_TIMEOUT = Duration.ofMinutes(10);
+    private static final String USERNAME_PASSWORD_URL_FORMAT = "{0}:{1}";
+    private static final int ERROR_RESPONSE_BODY_MAX_LENGTH = 4 * 1024;
+
+    private static final long WAIT_TIME_BETWEEN_ASYNC_JOB_UPDATES_IN_MILLIS = Duration.ofSeconds(3)
+                                                                                      .toMillis();
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(AsyncUploadJobExecutor.class);
+    private final ResilientOperationExecutor resilientOperationExecutor = getResilientOperationExecutor();
+
+    private final ExecutorService asyncFileUploadExecutor;
+    private final ExecutorService deployFromUrlExecutor;
+    private final ApplicationConfiguration applicationConfiguration;
+    private final AsyncUploadJobService asyncUploadJobService;
+    private final FileService fileService;
+    private final DescriptorParserFacadeFactory descriptorParserFactory;
+    private final HttpClient httpClient;
+
+    @Inject
+    public AsyncUploadJobExecutor(ExecutorService asyncFileUploadExecutor, ExecutorService deployFromUrlExecutor,
+                                  ApplicationConfiguration applicationConfiguration, AsyncUploadJobService asyncUploadJobService,
+                                  FileService fileService, DescriptorParserFacadeFactory descriptorParserFactory) {
+        this.asyncFileUploadExecutor = asyncFileUploadExecutor;
+        this.deployFromUrlExecutor = deployFromUrlExecutor;
+        this.applicationConfiguration = applicationConfiguration;
+        this.asyncUploadJobService = asyncUploadJobService;
+        this.fileService = fileService;
+        this.descriptorParserFactory = descriptorParserFactory;
+        httpClient = buildHttpClient();
```

Comment:
This class is still big and do multiple tasks. Can we consider extracting parts of this class related to build/perform http requests to external servers to download content into another class and this one to invoke the other one and use to upload downloaded content into db/oss?